Refactor connect edges -- copied Rust logic

bluenote10 opened the pull request:
@rowanwins I made a direct translation of the Rust version onto your branch. I copied over all test cases I had added in https://github.com/21re/rust-geo-booleanop/pull/10 and they worked immediately, so I'm quite happy ;).

These are the new/touched test cases: [test_cases.pdf](https://github.com/w8r/martinez/files/4240749/test_cases.pdf)

Note: The diff looks much bigger than necessary, because the target branch only merged in master before we had added the new generic test cases. If you can first update your branch to the latest master it should be easier to review.

w8r commented:
H Fabian, I think I can simplify a bit the usage of the new flag, if it's always opposite to `resultTransition`. I have also added you as a collaborator

rowanwins commented on src/index.js:
```diff
@@ -73,41 +73,22 @@ export default function boolean(subject, clipping, operation) {
-        points.push([points[0][0], points[0][1]]);
+  // Convert contours to polygons
+  const polygons = [];
+  for (let contour of contours) {
```
This threw errors for me running `npm run build`
````
(plugin buble) CompileError: Transforming for-of statements is not fully supported. Use `transforms: { forOf: false }` to skip transformation and disable this error, or `transforms: { dangerousForOf: true }` if you know what you're doing. (88:6)
````
It was simple enough to swap out with regular for loops which I think would be more normal.

rowanwins replied:
And similar in other places in `index.js` using `for (let foo of bar) {`

bluenote10 replied:
Good point, I've fixed the problem with `npm run build` now.

rowanwins commented on src/sweep_event.js:
```diff
@@ -65,20 +65,24 @@ export default class SweepEvent {
+     */
+    this.outputContourId = -1;
+
+    this.isExteriorRing = true;   // TODO: Looks unused, remove?
```
Yeah I think this is obselete

bluenote10 replied:
I looked into removing it, but it is still used for determining the input contour IDs, which in turn is used in a few places across the algorithm. I have a feeling that both can be removed, but it is a bit non-trivial to do so, and can still be done in a later PR.

rowanwins commented on src/sweep_event.js:
```diff
@@ -113,16 +117,25 @@ export default class SweepEvent {
+   */
+  inResult() {
+    return this.resultTransition !== 0;
+  }
```
Could possibly use a [getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get)
````
getter inResult() {
    return this.resultTransition !== 0;
  }
````
This would allow you to call `something.inResult` like a property rather than having to swap to a method.

This is probably a minor quibble

w8r replied:
I think it doesn't have to be calculated dynamically and can be turned into a flag.

bluenote10 replied:
Good idea with the getter => adapted.

@w8r Note that `inResult` is only needed in one place now (at the top of the `orderEvents` function), so the number of `inResult` calls isn't significant. Maybe the additional memory of storing the value might even slow things down.

bluenote10 commented:
I've addressed the comments and will merge this now into the main refactor-connect-edges branch. Feel free to take over from there, I was mainly trying to port the Rust logic and was hoping I can leave the JS details for the experts ;).
